Fix CC1Main setting trigger undefined behavior

bricknerb opened the pull request:
This is caused by setting a temporary lambda to a `llvm::function_ref`.
The fix is to assign the lambda into a variable that outlives the `llvm::function_ref`.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -146,13 +146,15 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
   // `carbon-busybox clang -cc1`, which results in an equivalent `clang_main`
   // call.
-  driver.CC1Main = [](llvm::SmallVectorImpl<const char*>& cc1_args) -> int {
+  std::function<int(llvm::SmallVectorImpl<const char*> & cc1_args)> cc1_main =
```
```suggestion
  auto cc1_main =
```

Does this need to explicitly be a std::function, or is it enough to store the lambda directly?

bricknerb replied:
Good point, changing to auto.

jonmeow reviewed (approved):
Nice catch!

chandlerc reviewed (approved):
Nice! Giving the LG so you can merge in your TZ.

bricknerb committed:
Fix CC1Main setting trigger undefined behavior
This is caused by setting a temporary lambda to a `llvm::function_ref`.
The fix is to assign the lambda into a variable that outlives the `llvm::function_ref`.
